### DISCUSSION TITLE
[apex] New rule for Apex code

### DISCUSSION THREAD

**Discussion by Puneet1726** [pmd/pmd#3025] [Q&A] [answered]
[apex].

I want to add a new rule to Apex code. Currently I am running PMD through CLI.
Any documentation for configuring new rules?

**Comment by adangel** [comment, marked as answer]
There are in general two ways to write own custom rules:
* Simple rules can be created using XPath expressions: https://pmd.github.io/latest/pmd_userdocs_extending_writing_xpath_rules.html The easiest way to get start is using the designer: https://pmd.github.io/latest/pmd_userdocs_extending_designer_reference.html
* More complex rules can be written using java: https://pmd.github.io/latest/pmd_userdocs_extending_writing_pmd_rules.html

Let us know, if/how we can improve the documentation or if you have further questions.
